packages/runtime/src/runtime.ts

```diff
@@ -736,141 +595,139 @@ export class Runtime {
             // https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
-            fromAccount.closeApp(txnParam.appId); // remove app from local state
-            throw error;
+            if (error instanceof RuntimeError && error.number === 1007) {
+              this.ctxOp.closeApp(txnParam.fromAccount.addr, txnParam.appId); // remove app from local state
+              clearErrorBool = true;
+              break;
+            } else {
+              throw error;
+            }
           }
 
-          fromAccount.closeApp(txnParam.appId); // remove app from local state
+          this.ctxOp.closeApp(txnParam.fromAccount.addr, txnParam.appId); // remove app from local state
+          break;
+        }
+        case TransactionType.DeleteSSC: {
+          const appParams = this.getApp(txnParam.appId);
+          this.run(appParams["approval-program"], ExecutionMode.STATEFUL);
+          this.ctxOp.deleteApp(txnParam.appId);
+          break;
+        }
+        case TransactionType.ModifyAsset: {
+          const asset = this.getAssetDef(txnParam.assetID);
+          if (asset.manager !== txnParam.fromAccount.addr) {
+            throw new RuntimeError(RUNTIME_ERRORS.ASA.MANAGER_ERROR, { address: asset.manager });
+          }
+          // modify asset in ctx.
+          this.ctxOp.modifyAsset(txnParam.assetID, txnParam.fields);
           break;
         }
         case TransactionType.FreezeAsset: {
           const asset = this.getAssetDef(txnParam.assetID);
           if (asset.freeze !== txnParam.fromAccount.addr) {
             throw new RuntimeError(RUNTIME_ERRORS.ASA.FREEZE_ERROR, { address: asset.freeze });
           }
+          this.ctxOp.freezeAsset(txnParam.assetID, txnParam.freezeTarget, txnParam.freezeState);
           break;
         }
         case TransactionType.RevokeAsset: {
           const asset = this.getAssetDef(txnParam.assetID);
           if (asset.clawback !== txnParam.fromAccount.addr) {
             throw new RuntimeError(RUNTIME_ERRORS.ASA.CLAWBACK_ERROR, { address: asset.clawback });
           }
+          this.ctxOp.revokeAsset(
+            txnParam.recipient, txnParam.assetID,
+            txnParam.revocationTarget, txnParam.amount
+          );
           break;
         }
-        case TransactionType.DestroyAsset:
-        case TransactionType.ModifyAsset: {
+        case TransactionType.DestroyAsset: {
           const asset = this.getAssetDef(txnParam.assetID);
           if (asset.manager !== txnParam.fromAccount.addr) {
             throw new RuntimeError(RUNTIME_ERRORS.ASA.MANAGER_ERROR, { address: asset.manager });
           }
+          this.ctxOp.destroyAsset(txnParam.assetID);
           break;
         }
       }
+    });
+    if (clearErrorBool) {
+      this.store = this.ctx.state;
+      throw new RuntimeError(RUNTIME_ERRORS.TEAL.REJECTED_BY_LOGIC);
     }
+  }
```

Comment:
Can we move `this.processTransactions()` to `Ctx` class (if we are operating entirely on `ctx` in this function) ?

Reply:
yes